File: LeetCode/three_sum.cpp
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        vector<int> numsWithoutDupli;
        numMap.insert({0, 0});
        for(int num: nums){
            if(numMap.find(num) == numMap.end()){
                numMap.insert({num, 0});
                numsWithoutDupli.push_back(num);
            }
            numMap[num]++;
        }
        if(numMap[0] > 2){
            vector<int> newList = {0, 0, 0};
            ans.push_back(newList);
        }
        if(numMap[0] > 0){
            for(int num: numsWithoutDupli){
                if(num >= 0){
                    break;
                }
                if(numMap.find(-1 * num) != numMap.end()){
                    vector<int> newList = {0, num, -1 * num};
                    ans.push_back(newList);
                }
            }
        }
        // numMap.erase(0);
        for(int i = 0; i < numsWithoutDupli.size(); i++){
            int num = numsWithoutDupli[i];
            if(listMap.find(num) != listMap.end()){
                cout << "a) i = " << i << endl;
                pushAns(num, listMap[num], ans);
            }
            for(int j = 0; j < i; j++){
                pushListMap(numsWithoutDupli[j], num);
            }
            if(numMap[num] > 1){
                int tmp = num * (-1) * 2;
                if(num > 0 && numMap.find(tmp) != numMap.end()){
                    vector<int> list = {num, num};
                    vector<vector<int>> lists =  {list};
                    cout << "b) i = " << i << endl;
                    pushAns(tmp, lists, ans);
                }
                pushListMap(num, num);
            }
        }
        // for(int i = 0; i < nums.size(); i++){
        //     if(numMap.find(nums[i]) != numMap.end()){
        //         pushAns(ans, numMap[nums[i]], nums[i]);
        //     }
        //     if(numSet.find(nums[i]) != numSet.end()){
        //         pushNumMap(nums[i], nums[i]);
        //         continue;
        //     }
        //     for(auto itr = numSet.begin(); itr != numSet.end(); itr++){
        //         pushNumMap(nums[i], *itr);
        //     }
        //     numSet.insert(nums[i]);
        // }
        return ans;
    }

private:
    // unordered_set<int> numSet;
    // unordered_map<int, vector<vector<int>>> numMap;
    unordered_map<int, int> numMap;
    unordered_map<int, vector<vector<int>>> listMap;
    void pushListMap(int num1, int num2){
        int key = -1 * (num1 + num2);
        if(listMap.find(key) != listMap.end()){
            listMap.insert({key, vector<vector<int>>()});
        }
        vector<int> pair = {num1, num2};
        listMap[key].push_back(pair);
    }

    void pushAns(int num, vector<vector<int>>& numLists, vector<vector<int>>& ans){
        for(vector<int>& list: numLists){
            list.push_back(num);
            ans.push_back(list);
        }
    }
    // void pushAns(vector<vector<int>>& ans, vector<vector<int>>& list, int num){
    //     for(vector<int>& numPair: numMap[num]){
    //         if(numPair.size() > 2){
    //             continue;
    //         }
    //         numPair.push_back(num);
    //         ans.push_back(numPair);
    //     }
    // }
    // void pushNumMap(int num1, int num2){
    //     int key = (-1) * (num1 + num2);
    //     if(numMap.find(key) == numMap.end()){
    //         numMap.insert({key, vector<vector<int>>()});
    //     }
    //     vector<int> newList = {num1, num2};
    //     numMap[key].push_back(newList);
    // }
};

int main(void){
    return 0;
}

File: LeetCode/three_sum.py
class Solution(object):
    def threeSum(self, nums):
        if len(nums) < 3:
            return []
        
        pos, neg = [], []
        count_zero = 0
        for num in nums:
            if num == 0:
                count_zero += 1
            elif num > 0:
                pos.append(num)
            else:
                neg.append(num)
        
        pos.sort()
        neg.sort()

        res = []
        if count_zero > 2:
            res.append([0, 0, 0])
        
        if count_zero > 0:
            self.twoSum(res, pos, neg)
        
        self.threeSumHalf(res, pos, neg)
        self.threeSumHalf(res, neg, pos)
        return res
    
    def twoSum(self, res, pos, neg):
        num_set = set()
        for num in pos:
            num_set.add(-num)
        
        for num in neg:
            if num in num_set:
                res.append([num, 0, -num])
                num_set.remove(num)
    
    def threeSumHalf(self, res, two_pick, one_pick):
        num_map = dict()
        num_set = set()
        for i in range(1, len(two_pick)):
            for j in range(0, i):
                key_str = str(two_pick[i]) + " " + str(two_pick[j])
                if key_str in num_set:
                    continue
                key = -two_pick[i] - two_pick[j]
                if key not in num_map:
                    num_map[key] = []
                num_map[key].append([two_pick[i], two_pick[j]])
                num_set.add(key_str)
        
        for num in one_pick:
            if num in num_map:
                for nums in num_map[num]:
                    nums.append(num)
                    res.append(nums)
                num_map.pop(num)

if __name__ == "__main__":
    nums = [-1, 0, 1, 2, -1, -4]
    sol = Solution()
    res = sol.threeSum(nums)
    print(res)